debugging-modules/README.md

```diff
@@ -0,0 +1,264 @@
+  Wasm-level information, and wraps a `WasmDebugger` instance.
+* **Wasm breakpoint:** A breakpoint associated with an instruction in the
+  debuggee module's code section.
+* **Source breakpoint:** A breakpoint associated with a location in the source
+  text, and may map to multiple Wasm breakpoints.
+
+Collectively, these concepts and their relationships can be visualized together
+in the following diagram:
+
+![](debugging-modules.png)
+
+## Walkthrough
+
+This section gives high-level walkthroughs of a few representative scenarios
+where debugging modules are used by developer tools to provide source-level
+information.
+
+### Symbolicating Profiler Stacks
+
+A sampling profiler will periodically pause Wasm execution, record its current
+stack, and then resume execution. Each frame in a recorded stack contains the
+Wasm module, function, and code offset. To display the profiler's results in a
+source-level presentation, the developer tools would perform the following steps
+for each frame of each stack:
+
+* Get or instantiate the debugger module `SourceDebugger` for the frame's Wasm
+  module.
+* Let `sourceRanges` be the result of calling `SourceDebugger.getSourceRanges`.
+* If `sourceRanges` is empty, leave the frame unsymbolicated.
+* Otherwise, let the frame's source-level symbolication be the first entry of
+  `sourceRanges`.
+
+> **Note:** This algorithm and usage of `SourceDebugger.getSourceRanges` enables
+> offline stack symbolication. It would not support symbolicating interpreted
+> code stacks like C# stacks within Blazor. If we wanted to support
+> symbolicating interpreted code stacks, it would likely require calling into
+> the debuggee module or inspecting its state online during stack sampling. In
+> turn, that would require only sampling at safe points of execution (similar to
+> GC safe points) to avoid making intermediate states or optimizations
+> observable. This can get a bit hairy, and whether we want to support this or
+> not is an open question.
+
+### Listing Sources and Displaying Source Text
+
+Debugger GUIs often display a list of source files or a file and directory tree
+of source files. To construct the complete list of source files for Wasm modules
+in a runtime, perform the following steps:
+
+* For each debuggee module in the runtime:
+    * Get or create the debugging module `SourceDebugger` for the debuggee
+      module.
+    * Append the result of calling `SourceDebugger.listSources` to the source
+      list.
+
+When a user selects a source from the list of all source files, a debugger GUI
+will typically display the selected source's text in a tab or panel. The text
+for a source can be retrieved by calling `SourceDebugger.getSourceText`.
+
+### Setting a Breakpoint
+
+When a user asks the debugger to set a source-level breakpoint, the debugger
+should perform the following steps:
+
+* Get the `SourceDebugger` associated with the breakpoint's source.
```

Comment:
I'd assume this should be a collection of `SourceDebugger`s if there is one per `Debugee Module`. For example, I can imagine common headers with inline functions (e.g. `<string>`), would result in the same source running in multiple modules.

Reply:
Great point! This should work out since even if we end up setting many breakpoints in the "same" source that end up in different debuggee modules, only one Wasm breakpoint is ever hit at a time.

I'll make an update in a little bit.